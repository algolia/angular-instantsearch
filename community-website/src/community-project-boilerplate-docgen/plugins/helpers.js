const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
