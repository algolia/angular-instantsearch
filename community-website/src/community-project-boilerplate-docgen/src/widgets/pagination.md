---
title: Widgets - Pagination
layout: widget.pug
canonical: https://www.algolia.com/doc/api-reference/widgets/pagination/angular/
---

# Pagination

## Description

The Pagination widget displays a simple pagination system allowing the user to change the current page.

## Options

| Attribute       | Type      | Description
| -               | -         | -
| `totalPages?`   | `number`  | Maximum number of pages to display
| `showFirst?`    | `boolean` | Display the first page link
| `showLast?`     | `boolean` | Display the last page link
| `showPrevious?` | `boolean` | Display the previous page link
| `showNext?`     | `boolean` | Display the next page link
| `padding?`      | `number`  | How many page links to display around the current page

## Code example

```html
<ais-pagination></ais-pagination>
```
