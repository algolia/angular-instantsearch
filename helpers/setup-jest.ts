import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
