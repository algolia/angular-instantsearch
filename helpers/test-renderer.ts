import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
