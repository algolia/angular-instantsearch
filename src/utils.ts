export function bem(widgetName: string) {
  const cx = function(element?: string, subElement?: string) {
    if (element) {
      const scoppedWidgetName = `ais-${widgetName}-${element}`;

      // output `ais-Widget-Header|Body|Footer ais-Header|Body|Footer`
      if (element === "header" || element === "body" || element === "footer") {
        const nonScoppedWidgetName = `ais-${element}`;
        return `${scoppedWidgetName} ${nonScoppedWidgetName}`;
      }

      // output `ais-Widget-Xyz--abc`
      if (subElement) {
        return `${scoppedWidgetName}--${subElement}`;
      }

      // output `ais-Widget-Xyz`
      return scoppedWidgetName;
    }

    // output `ais-Widget`
    return `ais-${widgetName}`;
  };
  return cx;
}

export function parseNumberInput(input?: number | string) {
  return typeof input === "string" ? parseInt(input, 10) : input;
}
