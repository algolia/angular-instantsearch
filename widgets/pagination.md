## `<ng-ais-pagination />`

The Pagination widget displays a simple pagination system allowing the user to change the current page.

| Attributes      | Type      | Description
| -               | -         | -
| `maxPages?`     | `number`  | Maximum number of pages to display
| `showFirst?`    | `boolean` | Display the first page link
| `showLast?`     | `boolean` | Display the last page link
| `showPrevious?` | `boolean` | Display the previous page link
| `showNext?`     | `boolean` | Display the next page link
| `pagesPadding?` | `number`  | How many page links to display around the current page
| `header?`       | `string`  | Displays text before widget
| `footer?`       | `string`  | Displays text after widget