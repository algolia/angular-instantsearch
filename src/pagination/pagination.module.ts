import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisPagination } from "./pagination";

@NgModule({
  declarations: [NgAisPagination],
  entryComponents: [NgAisPagination],
  exports: [NgAisPagination],
  imports: [CommonModule]
})
export class NgAisPaginationModule {}
