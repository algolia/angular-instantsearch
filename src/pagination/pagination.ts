import { Component, Input, Inject, PLATFORM_ID } from "@angular/core";
import { connectPagination } from "instantsearch.js/es/connectors";
import { noop, range } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { parseNumberInput } from "../utils";

@Component({
  selector: "ng-ais-pagination",
  template: `
    <div [class]="cx()">
      <ul [class]="cx('list')">
        <li
          *ngIf="showFirst"
          (click)="refine($event, 0)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'firstPage') +
            (state.currentRefinement === 0 ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(0)"
            [class]="cx('link')"
          >
            ‹‹
          </a>
        </li>

        <li
          *ngIf="showPrevious"
          (click)="refine($event, state.currentRefinement - 1)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'previousPage') +
            (state.currentRefinement === 0 ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.currentRefinement - 1)"
            [class]="cx('link')"
          >
            ‹
          </a>
        </li>

        <li
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'page') +
            (state.currentRefinement === page ? ' ' + cx('item', 'selected') : '')
          "
          *ngFor="let page of pages"
          (click)="refine($event, page)"
        >
          <a
            [class]="cx('link')"
            [href]="state.createURL(page)"
          >
            {{page + 1}}
          </a>
        </li>

        <li
          *ngIf="showNext"
          (click)="refine($event, state.currentRefinement + 1)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'nextPage') +
            (state.currentRefinement + 1 === state.nbPages ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.currentRefinement + 1)"
            [class]="cx('link')"
          >
            ›
          </a>
        </li>

        <li
          *ngIf="showLast"
          (click)="refine($event, state.nbPages)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'lastPage') +
            (state.currentRefinement + 1 === state.nbPages ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.nbPages)"
            [class]="cx('link')"
          >
            ››
          </a>
        </li>
      </ul>
    </div>
  `
})
export class NgAisPagination extends BaseWidget {
  // render options
  @Input() public showFirst: boolean = true;
  @Input() public showLast: boolean = false;
  @Input() public showPrevious: boolean = true;
  @Input() public showNext: boolean = true;
  @Input() public pagesPadding: number | string = 3;

  // connector optionsw
  @Input() public maxPages?: number | string;

  public state = {
    createURL: noop,
    currentRefinement: 0,
    nbHits: 0,
    nbPages: 0,
    refine: noop
  };

  get pages() {
    const { nbPages, currentRefinement } = this.state;

    const pagesArray = Array.apply(null, { length: nbPages }).map(
      Number.call,
      Number
    );

    const pagesPadding =
      typeof this.pagesPadding === "string"
        ? parseInt(this.pagesPadding, 10)
        : this.pagesPadding;

    if (pagesPadding && pagesPadding > 0) {
      // should not display pages that does not exists
      if (nbPages < pagesPadding * 2 + 1) {
        return pagesArray;
      }

      const minDelta = currentRefinement - pagesPadding - 1;
      const maxDelta = currentRefinement + pagesPadding + 1;

      if (minDelta < 0) {
        return range(0, currentRefinement + pagesPadding + Math.abs(minDelta));
      }

      if (maxDelta > nbPages) {
        return range(
          currentRefinement - pagesPadding - (maxDelta - nbPages),
          nbPages
        );
      }

      return range(
        currentRefinement - pagesPadding,
        currentRefinement + pagesPadding + 1
      );
    }

    return pagesArray;
  }

  constructor(
    @Inject(PLATFORM_ID) public platformId: Object,
    searchInstance: NgAisInstance
  ) {
    super(searchInstance, "Pagination");
  }

  public ngOnInit() {
    this.createWidget(connectPagination, {
      maxPages: parseNumberInput(this.maxPages)
    });
    super.ngOnInit();
  }

  public refine(event: MouseEvent, page: number) {
    event.stopPropagation();
    event.preventDefault();

    if (page <= this.state.nbPages && page >= 0) {
      this.state.refine(page);
    }
  }
}
