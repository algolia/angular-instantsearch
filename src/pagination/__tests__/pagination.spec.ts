import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisPagination } from "../pagination";

import { bem } from "../../utils";

const cx = bem("Pagination");

const defaultState = {
  createURL: jest.fn(),
  currentRefinement: 0,
  nbHits: 100,
  nbPages: 20,
  refine: jest.fn()
};

const render = createRenderer({
  defaultState,
  template: "<ng-ais-pagination></ng-ais-pagination>",
  TestedWidget: NgAisPagination
});

describe("Pagination", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders with pages in state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("renders does not uses pagesPadding when nbPages < pagesPadding * 2 + 1", () => {
    const fixture = render({ nbPages: 5 });
    const pages = fixture.debugElement.nativeElement.querySelectorAll(
      "." + cx("item", "page")
    );

    expect(pages.length).toBe(5);
    expect(fixture).toMatchSnapshot();
  });

  it("should refine when clicking a page", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const el = fixture.debugElement.nativeElement;
    const [, secondPage] = el.querySelectorAll("." + cx("item", "page"));
    secondPage.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(1);

    fixture.componentInstance.testedWidget.state.currentRefinement = 3;
    fixture.detectChanges();

    const previous = el.querySelector("." + cx("item", "previousPage"));
    previous.click();

    expect(refine).toHaveBeenLastCalledWith(2);

    const next = el.querySelector("." + cx("item", "nextPage"));
    next.click();

    expect(refine).toHaveBeenLastCalledWith(4);
  });

  it("should display last page button", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    fixture.componentInstance.testedWidget.showLast = true;
    fixture.detectChanges();

    const lastPage = fixture.debugElement.nativeElement.querySelector(
      "." + cx("item", "lastPage")
    );

    lastPage.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.nbPages);
  });
});
