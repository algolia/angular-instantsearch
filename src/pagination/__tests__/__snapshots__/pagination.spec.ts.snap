// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders does not uses pagesPadding when nbPages < pagesPadding * 2 + 1 1`] = `
<ng-ais-pagination
  cx={[Function Function]}
  pagesPadding={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showFirst={[Function Boolean]}
  showLast="false"
  showNext={[Function Boolean]}
  showPrevious={[Function Boolean]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Pagination"
  >
    
        
    <div
      class="ais-Pagination-body ais-body"
    >
      
          
      <ul
        class="ais-Pagination-list"
      >
        
            
        
        <li
          class="ais-Pagination-item--firstPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                «
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--previousPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ‹
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--page ais-Pagination-item--selected"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                1
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                2
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                3
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                4
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                5
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--nextPage"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ›
              
          </a>
          
            
        </li>
        
  
            
        
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-pagination>
`;

exports[`Pagination renders markup without state 1`] = `
<ng-ais-pagination
  cx={[Function Function]}
  pagesPadding={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showFirst={[Function Boolean]}
  showLast="false"
  showNext={[Function Boolean]}
  showPrevious={[Function Boolean]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div>
    
        
    <div>
      
          
      <ul>
        
            
        
        
  
            
        
        
  
            
        
        
  
            
        
        
  
            
        
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-pagination>
`;

exports[`Pagination renders with pages in state 1`] = `
<ng-ais-pagination
  cx={[Function Function]}
  pagesPadding={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showFirst={[Function Boolean]}
  showLast="false"
  showNext={[Function Boolean]}
  showPrevious={[Function Boolean]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Pagination"
  >
    
        
    <div
      class="ais-Pagination-body ais-body"
    >
      
          
      <ul
        class="ais-Pagination-list"
      >
        
            
        
        <li
          class="ais-Pagination-item--firstPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                «
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--previousPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ‹
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--page ais-Pagination-item--selected"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                1
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                2
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                3
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                4
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                5
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                6
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                7
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--nextPage"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ›
              
          </a>
          
            
        </li>
        
  
            
        
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-pagination>
`;

exports[`Pagination should display an header and a footer 1`] = `
<ng-ais-pagination
  cx={[Function Function]}
  footer={[Function String]}
  header={[Function String]}
  pagesPadding={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showFirst={[Function Boolean]}
  showLast="false"
  showNext={[Function Boolean]}
  showPrevious={[Function Boolean]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Pagination"
  >
    
        
    <div
      class="ais-Pagination-body ais-body"
    >
      
          
      <ul
        class="ais-Pagination-list"
      >
        
            
        
        <li
          class="ais-Pagination-item--firstPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                «
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--previousPage ais-Pagination-item--disabled"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ‹
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--page ais-Pagination-item--selected"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                1
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                2
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                3
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                4
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                5
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                6
              
          </a>
          
            
        </li>
        <li
          class="ais-Pagination-item--page"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                7
              
          </a>
          
            
        </li>
        
  
            
        
        <li
          class="ais-Pagination-item--nextPage"
        >
          
              
          <a
            class="ais-Pagination-link"
            href="null"
          >
            
                ›
              
          </a>
          
            
        </li>
        
  
            
        
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-pagination>
`;
